tests/structuredheadings/schemediscovery.js

```diff
@@ -0,0 +1,162 @@
+    },
+
+    tearDown: function () {
+      // reset the plugin between tests
+      var editor = this.editorBot.editor;
+      editor.plugins.structuredheadings.currentScheme = "1.1.1.1.1.";
+    },
+
+    "default is numbered": function () {
+      var bot = this.editorBot;
+      var editor = bot.editor;
+      bot.setHtmlWithSelection(
+        "<h1 class=\"autonumber autonumber-0\">^foo</h1>" +
+        "<h2 class=\"autonumber autonumber-1\">bar</h2>"
+      );
+
+      assert.areEqual(
+        "1.1.1.1.1.",
+        detectScheme(editor)
+      );
+    },
+
+    "invalid combos return numbered": function () {
+      var bot = this.editorBot;
+      var editor = bot.editor;
+      bot.setHtmlWithSelection(
+        "<h1 class=\"autonumber autonumber-0 autonumber-a\">^foo</h1>" +
+        "<h2 class=\"autonumber autonumber-1 autonumber-N\">bar</h2>"
+      );
+
+      assert.areEqual(
+        "1.1.1.1.1.",
+        detectScheme(editor)
+      );
+    },
+
+    "can detect based off of one level": function () {
+      var bot = this.editorBot;
+      var editor = bot.editor;
+      bot.setHtmlWithSelection(
+        "<h1 class=\"autonumber autonumber-0 autonumber-N\">^foo</h1>"
+      );
+
+      assert.areEqual(
+        "1. a. i. a. i.",
+        detectScheme(editor)
+      );
+    },
+
+    "autodetect will detect a default scheme document": function () {
+      var bot = this.editorBot;
+      var editor = bot.editor;
+
+      var listener = editor.on(
+        "dataReady",
+        function () {
+          listener.removeListener();
+          resume(function () {
+            assert.areEqual(
+              "1.1.1.1.1.",
+              getCurrentScheme(editor)
+            );
+          });
+        },
+        null,
+        null,
+        VERY_LAST_PRIORITY
+      );
+
+      editor.setData(
+        "<h1 class=\"autonumber autonumber-0\">^foo</h1>" +
+        "<h2 class=\"autonumber autonumber-1\">foo</h2>"
+      );
+
+      wait();
+    },
+
+    "autodetect falls to default if nothing matches": function () {
+      var bot = this.editorBot;
+      var editor = bot.editor;
+      var listener = editor.on(
+        "dataReady",
+        function () {
+          listener.removeListener();
+          resume(function () {
+            assert.areEqual(
+              "1.1.1.1.1.",
+              getCurrentScheme(editor)
+            );
+          });
+        },
+        null,
+        null,
+        VERY_LAST_PRIORITY
+      );
+
+      // A.A. is not a valid default
+      editor.setData(
+        "<h1 class=\"autonumber autonumber-0 autonumber-A\">^foo</h1>" +
+        "<h2 class=\"autonumber autonumber-1 autonumber-A\">foo</h2>"
+      );
+
+      wait();
+    },
+
+    "autodetect will detect a schemed document": function () {
+      var bot = this.editorBot;
+      var editor = bot.editor;
+
+      var listener = editor.on(
+        "dataReady",
+        function () {
+          listener.removeListener();
```

Comment:
editor cleanup kinda sucks, so the listener has to remove itself (this is a pattern used in their tests, without which I would have been stuck on this a lot longer)